apm-agent-plugins/apm-grpc/apm-grpc-plugin/src/main/java/co/elastic/apm/agent/grpc/ServerCallListenerInstrumentation.java

```diff
@@ -151,7 +148,14 @@ public FinalMethodCall(ElasticApmTracer tracer) {
                 //
                 // call complete (but client not guaranteed to get all messages)
                 // --> end of unary call (success)
-                .or(named("onComplete"));
+                .or(named("onComplete"))
+                //
+                // client completed all message sending, but can still cancel the call
+                // --> for unary calls, actual method invocation is done within 'onHalfClose' method, and there is no
+                // call to 'onComplete', thus consider it as 'final' allows to properly perform cleanup as it's the last
+                // method that will be invoked on the listener.
+                //
+                .or(named("onHalfClose"));
```

Comment:
this change was the one that fixed the memory leak, as it seems that some calls to `onComplete` are not always executed at all, or are executed in an unexpected order.

Reply:
I don't think that's the issue. The problem is that we're only instrumenting server call listeners, that contain `Unary` in the class name. The `UnaryServerCallListener` doesn't override the noop `onComplete` callback.

Another issue is that the transaction is ended in `co.elastic.apm.agent.grpc.ServerCallInstrumentation#onExit` (on `ServerCall#close`). The effect is that even when instrumenting `io.grpc.ServerCall.Listener#onComplete`, the helper is not called, because the transaction has already ended.

Furthermore, the fact that `inFlightServerListeners` is of type `WeakConcurrentMap<ServerCall.Listener<?>, ServerCall<?, ?>>` imposes a risk of a memory leak as `ServerCall` is not a weak reference. Have you considered mapping directly to the transaction (`WeakConcurrentMap<ServerCall.Listener<?>, Transaction>`)?

Similar story for client instrumentation. Changing the signature to `WeakConcurrentMap<ClientCall.Listener<?>, Span> inFlightClientListeners` would eliminate the chance of leaks and allows direct lookup of a span given a listener.

Another simplification of the client instrumentation could be done by removing `co.elastic.apm.agent.grpc.ClientCallImplInstrumentation.Constructor` and doing that inside `co.elastic.apm.agent.grpc.ChannelInstrumentation` (which instruments `io.grpc.Channel#newCall`). That would be more in line with the server instrumentation in `co.elastic.apm.agent.grpc.ServerCallHandlerInstrumentation` (instrumenting `io.grpc.ServerCallHandler#startCall`). And allows you to start the span before the constructor is invoked. In the exit advice you can then use the return value.